sys: heap: support maximum allocated bytes statistic

Damian-Nordic opened the pull request:
Besides the current allocated/free bytes, keep track of the maximum allocated bytes to help determine the heap size requirements. Also, provide a function to reset the statistic.

Signed-off-by: Damian Krolik <damian.krolik@nordicsemi.no>

dcpleung commented on include/sys/sys_heap.h:
```diff
@@ -82,6 +83,17 @@ struct sys_heap_runtime_stats {
+ * @param heap Pointer to sys_heap
+ * @return -EINVAL if null pointer was passed, otherwise 0
+ */
+int sys_heap_runtime_stats_reset_historical(struct sys_heap *heap);
```
Maybe `_update_` instead of `_reset_`? Reset, for some, might mean setting it back to zero.

Damian-Nordic replied:
Hmm. Isn't resetting more about setting back to the initial value rather than to zero? I thought that for the **maximum** kinds of statistics, the current value is a good initial value, but maybe it's just me. On the other hand, `_update_` may indicate many different things, so I would prefer to assign it a more descriptive name. What about something like `_remove_history` or `_history_restart`?

dcpleung replied:
hm... in this context, heap is initially empty so max allocated bytes would be zero. Maybe `_init_historical()`?

Damian-Nordic replied:
ok, I have done the change.

Damian-Nordic committed:
sys: heap: support maximum allocated bytes statistic

Besides the current allocated/free bytes, keep track of
the maximum allocated bytes to help determine the heap
size requirements. Also, provide a function to reset
the statistic.

Signed-off-by: Damian Krolik <damian.krolik@nordicsemi.no>

npitre commented:
I fail to see the purpose of `sys_heap_runtime_stats_init_historical()`.
What is it used for?

npitre commented:
OK. However `sys_heap_runtime_stats_init_historical()` is rather
suboptimal as a name for what it does. Maybe
`sys_heap_runtime_stats_reset_max()` would be more self-descriptive?

Damian-Nordic commented:
> OK. However `sys_heap_runtime_stats_init_historical()` is rather suboptimal as a name for what it does. Maybe `sys_heap_runtime_stats_reset_max()` would be more self-descriptive?

I changed the original name `_reset_historical`  to `_init_historical` because the former was contested by @dcpleung. Daniel, would you accept `_reset_max` or should we keep searching? :)

Damian-Nordic commented:
Ok, renamed to `sys_heap_runtime_stats_reset_max`.
